### PR TITLE
New java manual

### PR DESCRIPTION
**Pull request by stefano-ottolenghi**

### CODE REVIEW

**Comment by neo-technology-commit-status-publisher** [comment]
Looks like you've updated the documentation!

Check out your changes at https://neo4j-docs-drivers-413.surge.sh

File: java-manual/modules/ROOT/pages/connect.adoc [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+----
+package main;
+
+import org.neo4j.driver.AuthTokens;
+import org.neo4j.driver.GraphDatabase;
+
+public class App {
+
+    // URI examples: "neo4j://localhost", "neo4j+s://xxx.databases.neo4j.io"
+    private static final String URI = "<URI for Neo4j database>";
```
Comment (michael-simons) [inline java-manual/modules/ROOT/pages/connect.adoc]: Unsure if I would use constants in the connect example, while they are helpful to indicate what the parameter does by their name, they otherwise clutter things.

File: java-manual/modules/ROOT/pages/connect.adoc [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+    }
+}
+----
+
+<1> Creating a `Driver` instance only provides information on _how_ to access the database, but does not actually _establish_ a connection.
+Connection is instead deferred to when the first query is executed.
+<2> To verify immediately that the driver can connect to the database (valid credentials, compatible versions, etc), use the `.verifyConnectivity()` method after initializing the driver.
+
+Both the creation of a `Driver` object and the connection verification can raise a number of different link:https://neo4j.com/docs/api/java-driver/current/org.neo4j.driver/org/neo4j/driver/exceptions/package-summary.html[exceptions].
+Since a connection error is a blocker for any subsequent task, the most common choice is to let the program crash should an exception occur during connection.
```
Comment (michael-simons) [inline java-manual/modules/ROOT/pages/connect.adoc]: That's bold but true!

Please rephrase:
> should an exception occur during connection.

into 

> should an exception occur while establishing a connection.

File: java-manual/modules/ROOT/pages/connect.adoc [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+
+`Driver` objects are _immutable_, _thread-safe_, and fairly _expensive to create_.
+Share them across threads (but not across processes) and use xref:query-simple#impersonation[impersonation] to query the database with a different user.
+If you want to alter a `Driver` configuration, you will need to create a new object.
+
+
+== Connect to an Aura instance
```
Comment (michael-simons) [inline java-manual/modules/ROOT/pages/connect.adoc:46]: Please add some sentence that an Aura instance is conceptional not different than any other one, but deliver its credentials in said file.

File: java-manual/modules/ROOT/pages/connect.adoc [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+
+When you create an <<Aura>> instance, you may download a text file containing the connection information to the database.
+The file has a name of the form `Neo4j-a0a2fa1d-Created-2023-11-06.txt`.
+
+To connect to such an instance, you may either use the URI and authentication credentials explicitly in your application, or load the content of the connection file in the environment with `Dotenv.load()` and populate your local variables via `dotenv.get()`.
```
Comment (michael-simons) [inline java-manual/modules/ROOT/pages/connect.adoc]:
```suggestion
You can either extract the URI and the credentials from that text file manually or use a third party-module to load it. 
We recommend link:https://github.com/cdimascio/dotenv-java/[`dotenv-java`] for that purpose.
It provides a fluent API to use any Dotenv file in your Java program and access its content.
An example is shown below.
```

File: java-manual/modules/ROOT/pages/connect.adoc [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+
+`Driver` objects are _immutable_, _thread-safe_, and fairly _expensive to create_.
+Share them across threads (but not across processes) and use xref:query-simple#impersonation[impersonation] to query the database with a different user.
+If you want to alter a `Driver` configuration, you will need to create a new object.
+
+
+== Connect to an Aura instance
+
+When you create an <<Aura>> instance, you may download a text file containing the connection information to the database.
+The file has a name of the form `Neo4j-a0a2fa1d-Created-2023-11-06.txt`.
```
Comment (michael-simons) [inline java-manual/modules/ROOT/pages/connect.adoc:49]:
```suggestion
The file has a name of the form `Neo4j-a0a2fa1d-Created-2023-11-06.txt`.
It is a so-called Dotenv file, containing various environment variables for a process to work.
```

File: java-manual/modules/ROOT/pages/connect.adoc [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+            .filename("Neo4j-a0a2fa1d-Created-2023-11-06.txt")
+            .load();
+
+        final String URI = dotenv.get("NEO4J_URI");
+        final String USER = dotenv.get("NEO4J_USERNAME");
+        final String PASSWORD = dotenv.get("NEO4J_PASSWORD");
+
+        try (var driver = GraphDatabase.driver(URI, AuthTokens.basic(USER, PASSWORD))) {
+            driver.verifyConnectivity();
+        }
```
Comment (michael-simons) [inline java-manual/modules/ROOT/pages/connect.adoc]:
```suggestion
        var dotenv = Dotenv.configure()
            .directory("/path/to/env/file")
            .filename("Neo4j-a0a2fa1d-Created-2023-11-06.txt")
            .load();

        var uri = dotenv.get("NEO4J_URI");
        var user = dotenv.get("NEO4J_USERNAME");
        var password = dotenv.get("NEO4J_PASSWORD");

        try (var driver = GraphDatabase.driver(uri, AuthTokens.basic(user, password))) {
            driver.verifyConnectivity();
        }
```

UPPERCASE only for class constants, never for variables, even if they are final.

File: java-manual/modules/ROOT/pages/connect.adoc [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+    }
+}
+----
+
+
+== Close connections
+
+Always close `Driver` objects to free up all allocated resources, even upon unsuccessful connection or runtime errors.
+Either instantiate the `Driver` object as a resource using the `try` construct, or call the `Driver.close()` method explicitly.
```
Comment (michael-simons) [inline java-manual/modules/ROOT/pages/connect.adoc]:
```suggestion
Either use the `Driver` object as a resource inside a try-with-resources statement as shown in our examples, or call the `Driver.close()` method explicitly.
```

File: java-manual/modules/ROOT/pages/query-simple.adoc [hunk 8]
```diff
@@ -0,0 +1,363 @@
+= Query the database
+
+Once you have xref:connect.adoc[connected to the database], you can run queries using <<Cypher>> and the method `Driver.executableQuery()`.
```
Comment (michael-simons) [inline java-manual/modules/ROOT/pages/query-simple.adoc]:
```suggestion
Once you have xref:connect.adoc[connected to the database], you can run execute <<Cypher>> queries using via `Driver.executableQuery()`.
```

File: java-manual/modules/ROOT/pages/query-simple.adoc [hunk 9]
```diff
@@ -0,0 +1,363 @@
...
+
+
+== Write to the database
+
+To create a node representing a person named `Alice`, use the Cypher clause link:{neo4j-docs-base-uri}/cypher-manual/current/clauses/merge/[`MERGE`]:
+
+.Create a node representing a person named `Alice`
+[source, java]
+----
+// import org.neo4j.driver.QueryConfig;
```
Comment (michael-simons) [inline java-manual/modules/ROOT/pages/query-simple.adoc:19]: Probably a left over?

File: java-manual/modules/ROOT/pages/query-simple.adoc [hunk 10]
```diff
@@ -0,0 +1,363 @@
...
+
+var result = driver.executableQuery("MERGE (:Person {name: $name})")  // <1>
+    .withParameters(Map.of("name", "Alice"))  // <2>
```
Comment (michael-simons) [inline java-manual/modules/ROOT/pages/query-simple.adoc:22]: I would show the required imports, actually the whole app (including the try block…)
At least people are gonna wonder which `Map` class this is and in the next line, from where the `QueryConfig` comes from.

File: java-manual/modules/ROOT/pages/query-simple.adoc [hunk 11]
```diff
@@ -0,0 +1,363 @@
...
+    .withConfig(QueryConfig.builder().withDatabase("neo4j").build())  // clunky  <3>
+    .execute();
+var summary = result.summary();
+System.out.printf("Created %s nodes in %d ms.\n",
```
Comment (michael-simons) [inline java-manual/modules/ROOT/pages/query-simple.adoc]:
```suggestion
System.out.printf("Created %s nodes in %d ms.%n",
```

I do find the usage of `printf` unusual. Most people would use `println` and than just string concatenate I guess.
If you do `printf`, use at least `%n`, which gives you the system specific new line (sometimes even \n\r).

File: java-manual/modules/ROOT/pages/query-simple.adoc [hunk 12]
```diff
@@ -0,0 +1,363 @@
...
+----
+
+<1> specifies the Cypher query
+<2> is a map of _query parameters_
+<3> specifies which database the query should be run against
+
+[NOTE]
+`MERGE` creates a new node matching the requirements unless one already exists.
+If a matching node already exists, it is returned.
+For _strict creation_, use the Cypher clause link:{neo4j-docs-base-uri}/cypher-manual/current/clauses/create/[`CREATE`].
```
Comment (michael-simons) [inline java-manual/modules/ROOT/pages/query-simple.adoc]: Read "strict creation" here for the first time.

I would probably just use `CREATE`, really. `MERGE` is complicated and for things to work proper, there must be constraints.

File: java-manual/modules/ROOT/pages/query-simple.adoc [hunk 13]
```diff
@@ -0,0 +1,363 @@
...
+
+// Loop through results and do something with them
+var records = result.records()  // <1>
+records.forEach(r -> {
+    System.out.println(r);  // or r.get("name").asString()
+});
+
+// Summary information
+var summary = result.summary();  // <2>
+System.out.printf("The query %s returned %d records in %d ms.\n",
```
Comment (michael-simons) [inline java-manual/modules/ROOT/pages/query-simple.adoc]:
```suggestion
System.out.printf("The query %s returned %d records in %d ms.%n",
```

File: java-manual/modules/ROOT/pages/query-simple.adoc [hunk 14]
```diff
@@ -0,0 +1,363 @@
...
+    .withParameters(Map.of("name", "Alice", "friend", "Bob"))
+    .execute();
+var summary = result.summary();
+System.out.println("Query updated the database?");
+System.out.println(summary.counters().containsUpdates());
+----
+
+[WARNING]
+====
+It might feel tempting to create new relationships with a single `MERGE` clause, such as: +
```
Comment (michael-simons) [inline java-manual/modules/ROOT/pages/query-simple.adoc]: This box should move upwards I think and include a bit about constraints. IMHO we are leaving driver land here though and going Cypher. I would probably spare me the hassle and focus on executing Cypher alone.

File: java-manual/modules/ROOT/pages/query-simple.adoc [hunk 15]
```diff
@@ -0,0 +1,363 @@
...
+    .withConfig(QueryConfig.builder().withDatabase("neo4j").build())
+    .execute();
+----
+
+[NOTE]
+There can be circumstances where your query structure prevents the usage of parameters in all its parts.
+For those advanced use cases, see xref:query-advanced#_dynamic_values_in_property_keys_relationship_types_and_labels[Dynamic values in property keys, relationship types, and labels].
+
+
+== Error handling
```
Comment (michael-simons) [inline java-manual/modules/ROOT/pages/query-simple.adoc:171]: This feels very odd to me. This is the case for all methods in all programs. 

I would rewrite this something along the following idea:


Running a query against a remote server can fail for a number of reasons. Some of them are due to syntactically wrong queries and similar, client side issues, some of them can be caused by network issues. The driver will retry running your query in many scenarios, such as temporary server unavailability. 
For all client side exceptions it is up to you if program defensive (i.e. check if records are available to avoid `[NoSuchRecordException](https://neo4j.com/docs/api/java-driver/current/org.neo4j.driver/org/neo4j/driver/exceptions/NoSuchRecordException.html)`), or if you just catch and handle them).

File: java-manual/modules/ROOT/pages/query-simple.adoc [hunk 16]
```diff
@@ -0,0 +1,363 @@
...
+----
+try {
+    driver.executableQuery(...).execute()
+} catch (Exception as e) {
+    ...  // handle exception
+}
+----
+
+[TIP]
+The driver automatically retries to run a failed query, if the failure is deemed to be transient (for example due to temporary server unavailability).
```
Comment (michael-simons) [inline java-manual/modules/ROOT/pages/query-simple.adoc]: Remove this if you follow my above suggestion.

File: java-manual/modules/ROOT/pages/query-simple.adoc [hunk 17]
```diff
@@ -0,0 +1,363 @@
...
+
+
+== Query configuration
+
+You can supply further configuration parameters to alter the default behavior of `.executableQuery()`.
+You do so through the method `.withConfig()`, which takes a link:https://neo4j.com/docs/api/java-driver/current/org.neo4j.driver/org/neo4j/driver/QueryConfig.html[`QueryConfig`] object.
+
+
+=== Database selection
```
Comment (michael-simons) [inline java-manual/modules/ROOT/pages/query-simple.adoc:187]: Just a general remark: I find this absolutely terrible and am wishing for quite some time that if no database is set on the driver, the drivers just should try `neo4j` and be done with it.

The examples look horrible with that dance and feel very much "in your face".

File: java-manual/modules/ROOT/pages/query-simple.adoc [hunk 18]
```diff
@@ -0,0 +1,363 @@
...
+
+var result = driver.executableQuery("MATCH (p:Person) RETURN p.name")
+    .withConfig(QueryConfig.builder()
+        .withDatabase("neo4j")
+        .withRouting(RoutingControl.READ)  // mark-line
+        .build())
+    .execute();
+----
+
+Even though routing a _write_ query to _read_ nodes will likely result in a runtime error, xref:transactions#access-control-note[do not rely on this for security purposes].
```
Comment (michael-simons) [inline java-manual/modules/ROOT/pages/query-simple.adoc]:
```suggestion
The drivers do not parse your queries.
If you sent a _write_ query, such as a `CREATE (:Movie)` statement, they will just pass this on, preferring read nodes, which will eventually lead to a runtime exception.
Specifying the routing control is only about picking primary or secondary servers and not about xref:transactions#access-control-note[access control]. 
It must not be used for security purposes.
```

File: java-manual/modules/ROOT/pages/query-simple.adoc [hunk 19]
```diff
@@ -0,0 +1,363 @@
...
+
+== A full example
+
+[source, java]
+----
+package demo;
+
+import java.util.Map;
+import java.util.List;
+import java.util.Arrays;
```
Comment (michael-simons) [inline java-manual/modules/ROOT/pages/query-simple.adoc]:
```suggestion
```

File: java-manual/modules/ROOT/pages/query-simple.adoc [hunk 20]
```diff
@@ -0,0 +1,363 @@
...
+
+    private static final String URI = "<URI for Neo4j database>";
+    private static final String USER = "<Username>";
+    private static final String PASSWORD = "<Password>";
+
+    public static void main(String... args) {
+
+        try (var driver = GraphDatabase.driver(URI, AuthTokens.basic(USER, PASSWORD))) {
+
+            List<Map> people = Arrays.asList(
```
Comment (michael-simons) [inline java-manual/modules/ROOT/pages/query-simple.adoc]:
```suggestion
            List<Map> people = List.of(
```

File: java-manual/modules/ROOT/pages/query-simple.adoc [hunk 21]
```diff
@@ -0,0 +1,363 @@
...
+                Map.of("name", "Bob", "age", 19),
+                Map.of("name", "Peter", "age", 50),
+                Map.of("name", "Anna", "age", 30)
+            );
+
+            try {
+
+                //Create some nodes
+                people.forEach(person -> {
+                    var result = driver.executableQuery("MEGE (p:Person {name: $person.name, age: $person.age})")
```
Comment (michael-simons) [inline java-manual/modules/ROOT/pages/query-simple.adoc]:
```suggestion
                    var result = driver.executableQuery("MERGE (p:Person {name: $person.name, age: $person.age})")
```

File: java-manual/modules/ROOT/pages/query-simple.adoc [hunk 22]
```diff
@@ -0,0 +1,363 @@
...
+                        .withConfig(QueryConfig.builder().withDatabase("neo4j").build())
```
Comment (michael-simons) [inline java-manual/modules/ROOT/pages/query-simple.adoc:292]: Store the config in a variable and reuse it.

File: java-manual/modules/ROOT/pages/query-simple.adoc [hunk 23]
```diff
@@ -0,0 +1,363 @@
...
+                    .execute();
+
+                // Loop through results and do something with them
+                result.records().forEach(r -> {
+                    System.out.println(r);  // doesn't unwrap properties, Record<{friend: node<4>}>, nor .fields() does
+                    // python shows <Record friend=<Node element_id='4:5bad7cf2-d17e-4d07-9266-b88c8720ebd2:4' labels=frozenset({'Person'}) properties={'name': 'Bob', 'age': 19}>>
+                });
+
+                // Summary information
+                System.out.printf("The query %s returned %d records in %d ms.",
```
Comment (michael-simons) [inline java-manual/modules/ROOT/pages/query-simple.adoc]:
```suggestion
                System.out.printf("The query %s returned %d records in %d ms.%n",
```

File: java-manual/modules/ROOT/pages/transactions.adoc [hunk 24]
```diff
@@ -0,0 +1,487 @@
...
+
+.Retrieve people whose name starts with `Al`.
+[source, java]
+----
+// import java.util.Map
+// import org.neo4j.driver.SessionConfig
+
+try (var session = driver.session(SessionConfig.builder().withDatabase("neo4j").build())) {  // <1>
+    var people = session.executeRead(tx -> {  // <2>
+        var result = tx.run("""
```
Comment (michael-simons) [inline java-manual/modules/ROOT/pages/transactions.adoc:50]: This example shows only one statement either. There is no difference towards the executableQuery variant.

File: java-manual/modules/ROOT/pages/transactions.adoc [hunk 25]
```diff
@@ -0,0 +1,487 @@
...
+        // use tx.run() to run queries
+        //     tx.commit() to commit the transaction
+        //     tx.rollback() to rollback the transaction
+    }
+}
+----
+
+An explicit transaction behaves as follows:
+
+- it gets committed automatically upon successful return, or manually through the method  link:https://neo4j.com/docs/api/java-driver/current/org.neo4j.driver/org/neo4j/driver/Transaction.html#commit()[`Transaction.commit()`]
```
Comment (michael-simons) [inline java-manual/modules/ROOT/pages/transactions.adoc]: *This is not true.*

You *must* use `commit` prior to close.

See JavaDoc of transaction

```
Transactions are typically wrapped in a try-with-resources block
which ensures in case of any error in try block, the transaction is
automatically rolled back on close. Note that <code>ROLLBACK</code> is the
default action unless {@link #commit()} is called before closing.
```

File: java-manual/modules/ROOT/pages/transactions.adoc [hunk 26]
```diff
@@ -0,0 +1,487 @@
...
+
+                try {
+                    decreaseCustomerBalance(tx, customerId, amount);
+                    tx.commit();
+                    System.out.printf("Transferred %f to %s.\n", amount, customerId);
+                } catch (Exception e) {
+                    requestInspection(customerId, otherBankId, amount, e);
+                    throw new RuntimeException(e.getMessage());
+                } finally {
+                    tx.close();  // rolls back if not yet committed
```
Comment (michael-simons) [inline java-manual/modules/ROOT/pages/transactions.adoc]: Superflous, it is already wrapper in try with resources. The `finally` block should go.

File: java-manual/modules/ROOT/pages/transactions.adoc [hunk 27]
```diff
@@ -0,0 +1,487 @@
...
+                }
+
+                otherBankTransferApi(customerId, otherBankId, amount);
+                // Now the money has been transferred => can't rollback anymore
+                // (cannot rollback external services interactions)
+
+                try {
+                    decreaseCustomerBalance(tx, customerId, amount);
+                    tx.commit();
+                    System.out.printf("Transferred %f to %s.\n", amount, customerId);
```
Comment (michael-simons) [inline java-manual/modules/ROOT/pages/transactions.adoc]:
```suggestion
                    System.out.printf("Transferred %f to %s.%n", amount, customerId);
```

File: java-manual/modules/ROOT/pages/transactions.adoc [hunk 28]
```diff
@@ -0,0 +1,487 @@
...
+    );
+    people.forEach(person -> System.out.println(person));
+}
+----
+
+
+== Close sessions
+
+Each connection pool has *a finite number of sessions*, so if you open sessions without ever closing them, your application could run out of them.
+It is thus recommended to create sessions using the `with` statement, which automatically closes them when the application is done with them.
```
Comment (michael-simons) [inline java-manual/modules/ROOT/pages/transactions.adoc]: `with` seems to be a Python left-over.

File: java-manual/modules/ROOT/pages/transactions.adoc [hunk 29]
```diff
@@ -0,0 +1,487 @@
...
+    .withDatabase("neo4j")
+    .withRouting(RoutingControl.READ)  // mark-line
+    .build());
+----
+
+[#access-control-note]
+// .Routing ≠ Access control
+[NOTE]
+====
+Although executing a _write_ query in read mode likely results in a runtime error, *you should not rely on this for access control.* The difference between the two modes is that _read_ transactions will be routed to any node of a cluster, whereas _write_ ones will be directed to the leader.
```
Comment (michael-simons) [inline java-manual/modules/ROOT/pages/transactions.adoc:432]: Same as in the previous file.
Be explicit that this is about routing only, use two sentences and "must not"


Also Neo4j 5 notation is "primary" and "secondary", afaik we don't talk leader/follower anymore.

**Review by michael-simons — CHANGES_REQUESTED** [review]

File: java-manual/modules/ROOT/pages/connect.adoc (hunk 1 above)
Comment (stefano-ottolenghi) [reply java-manual/modules/ROOT/pages/connect.adoc]: You mean local variables inside `main` would be more appropriate? How so? I'm interested in understanding why you wouldn't use class constants in your own app.

File: java-manual/modules/ROOT/pages/query-simple.adoc (hunk 9 above)
Comment (stefano-ottolenghi) [reply java-manual/modules/ROOT/pages/query-simple.adoc:19]: No, it does use `QueryConfig` to set the database.

File: java-manual/modules/ROOT/pages/query-simple.adoc (hunk 10 above)
Comment (stefano-ottolenghi) [reply java-manual/modules/ROOT/pages/query-simple.adoc:22]: True, I should also import `java.util.Map`. My intention is to keep examples short and to the point, to avoid cluttering each one with all the boilerplate of class and full imports, but showing "partial/new" imports as comments. If people need a full skeleton to work with, the `Full example` at the bottom provides it.

File: java-manual/modules/ROOT/pages/query-simple.adoc (hunk 11 above)
Comment (stefano-ottolenghi) [reply java-manual/modules/ROOT/pages/query-simple.adoc]: > Most people would use println and than just string concatenate I guess.

If that's what the java world does, let's do that here as well. Will address it so.

File: java-manual/modules/ROOT/pages/query-simple.adoc (hunk 12 above)
Comment (stefano-ottolenghi) [reply java-manual/modules/ROOT/pages/query-simple.adoc]: This is a choice I took more than a year ago, and in retrospect I think you're right. I'm hiding a share of complexity here for no reason, and possibly encouraging bad practice. Let's go with `CREATE`, and also in other driver manuals.

File: java-manual/modules/ROOT/pages/query-simple.adoc (hunk 14 above)
Comment (stefano-ottolenghi) [reply java-manual/modules/ROOT/pages/query-simple.adoc]: Yeah, it's a fine line between giving you a minimal cypher crash course if you're building an app without being proficient in cypher already (which I think you should _not_), and giving you too much detail. At the same time, I want to put you on guard for very natural mistakes like this one, but not go on a tangent. 

However, if we replace `MERGE` with `CREATE`, the whole problem vanishes and this note is not needed anymore.

File: java-manual/modules/ROOT/pages/query-simple.adoc (hunk 15 above)
Comment (stefano-ottolenghi) [reply java-manual/modules/ROOT/pages/query-simple.adoc:171]: I agree with you that I find this section border-line confusing. The reason it is here is user interviews from a few years back, where people were lamenting not knowing how to do "proper error handling", which I interpreted as "should I differentiate between the different exceptions the driver may raise, or how should I do?". But maybe I should have just ignored that part of user confusion since, as you say, this is how error handling in _any_ software works. We'd show it in the full examples anyway.

File: java-manual/modules/ROOT/pages/query-simple.adoc (hunk 17 above)
Comment (stefano-ottolenghi) [reply java-manual/modules/ROOT/pages/query-simple.adoc:187]: Agreed, but what can we do. 

They also look horrible because the query config methods are defined on `QueryConfig` rather than directly on `ExecutableQuery`. If it looked like this, it wouldn't be as bad:
```java
var result = driver.executableQuery("MATCH (p:Person {name: $name})")
    .withParameters(Map.of("name", "Alice"))
    .withDatabase("neo4j")
    .execute();
```
I planned to bring this up somewhere else. All this config ballet (also in sessions and drivers) looks like madness to me.

File: java-manual/modules/ROOT/pages/transactions.adoc (hunk 24 above)
Comment (stefano-ottolenghi) [reply java-manual/modules/ROOT/pages/transactions.adoc:50]: This is very true. I didn't want to make it into too a complicated example, since it already has 4 callouts, but I get your point that it doesn't really display a _transaction_. At this stage, the intention is mostly to teach you how to use transaction functions, and then the next example will show real transactions. Do you think this is a poor intention? Would it make it better to have a code comment after the first query saying smth like "run more queries"?

File: java-manual/modules/ROOT/pages/transactions.adoc (hunk 25 above)
Comment (stefano-ottolenghi) [reply java-manual/modules/ROOT/pages/transactions.adoc]: Ah, thanks for bringing this up. I was unsure but gave it for granted at this stage since this is how Python, Javascript, and Go work: they all default to commit if no error occurs. Interesting that Java took the opposite turn.

File: java-manual/modules/ROOT/pages/connect.adoc (hunk 1 above)
Comment (michael-simons) [reply java-manual/modules/ROOT/pages/connect.adoc]: The values that you store in those constants are not values that are usually constant, quite the contrary. You get them as user input, from Dotenv (which you do at the next step) or other config mechanism. They are conceptional no constants.

Having them as variables is helpful, as Java has no named parameters and so you see their meaning immediately.

Comment (michael-simons) [reply java-manual/modules/ROOT/pages/connect.adoc]: That said, I do use class constants a lot, if appropriate, see for example: https://github.com/michael-simons/neo4j-migrations/blob/main/neo4j-migrations-core/src/main/java/ac/simons/neo4j/migrations/core/Defaults.java

But take note: This class cannot be inherited, cannot be instantiated. I care about these semantics a lot.

File: java-manual/modules/ROOT/pages/transactions.adoc (hunk 25 above)
Comment (michael-simons) [reply java-manual/modules/ROOT/pages/transactions.adoc]: Is this true? If so you, you should raise this with our colleagues. 

They will be in for a surprise wrt GQL: GQL requires exactly the behaviour of the Java driver, there will be no auto-commit.

Comment (robsdedude) [reply java-manual/modules/ROOT/pages/transactions.adoc]: I checked the Python manual, and it does not specify that explicit transactions are automatically committed. And that's good, because they aren't.

This is what it says:
> Closing an explicit transaction can either happen automatically at the end of a with block, or can be explicitly controlled through the methods [Transaction.commit()](https://neo4j.com/docs/api/python-driver/current/api.html#neo4j.Transaction.commit), [Transaction.rollback()](https://neo4j.com/docs/api/python-driver/current/api.html#neo4j.Transaction.rollback), or [Transaction.close()](https://neo4j.com/docs/api/python-driver/current/api.html#neo4j.Transaction.close).

Admittedly, it's not super clear about what happens either.

Comment (stefano-ottolenghi) [reply java-manual/modules/ROOT/pages/transactions.adoc]: Weeeeell ... to me any return within a `with` block counts as reaching the end of the `with` block :grimacing:

Comment (stefano-ottolenghi) [reply java-manual/modules/ROOT/pages/transactions.adoc]: Either way, if the default behavior at the end of a `with` block is to commit, that's definitely different than in java, where the default behavior at the end of a `try` block (which is the same as `with` in python) is to rollback.

Comment (robsdedude) [reply java-manual/modules/ROOT/pages/transactions.adoc]: No sorry. I might have expressed myself ambiguously. The Python driver will close transactions when leaving the with block. Closing a transaction that has not been committed or rolled back yet means rolling it back.

TL;DR: explicit transactions are never committed unless the user does so explicitly (`tx.commit()`).

With "not super clear", I mean it just talks about "closing", without further specifying what that means.

Comment (stefano-ottolenghi) [reply java-manual/modules/ROOT/pages/transactions.adoc]: Thanks for clarifying -- I misunderstood the driver's behavior 1.5 years ago, and now read in your message what I wanted to read :grimacing: I'll also clarify in other manuals then.

**Comment by neo-technology-commit-status-publisher** [comment]
Thanks for the documentation updates.

The preview documentation has now been torn down - reopening this PR will republish it.

File: java-manual/modules/ROOT/pages/transactions.adoc (hunk 24 above)
Comment (stefano-ottolenghi) [reply java-manual/modules/ROOT/pages/transactions.adoc:50]: My compromise on this is
- leave the example with one query as is, but add a comment saying more queries would get executed in the same tx. I'd rather have the example a bit more stupid but where the reader can entirely focus on the driver API, as meaningful transaction examples are much longer (there's one just after)
- in the quickstart page, put the long hard transaction example.

**Review by michael-simons — CHANGES_REQUESTED** [review]
I sent you a whole list over on slack.
Most pressing items are the wrong version number on dependency declaration as well as the part on the `USE` (which affects the other manual, too).

The rest of my remarks are of smaller nature.

In general, this is a stellar overhaul, well done!

**Review by michael-simons — APPROVED** [review]
Thanks a lot for the great overhaul of the manual.
